tools/adios2pio-nm/adios2pio-nm-lib.cxx

```diff
@@ -1525,229 +1569,337 @@ int ConvertVariableTimedPutVar(IOVector &bpIO, EngineVector &bpReader,
-                                 int nblocks_per_step, int iosysid,
-                                 MPI_Comm comm, int mpirank, int nproc)
+                                 int ncid,
+                                 Variable &var,
+                                 DecompositionMap &decomp_map,
+                                 int iosysid,
+                                 std::string file0,
+                                 adios2::ADIOS &adios,
+                                 uint64_t time_step,
+                                 MPI_Comm comm,
+                                 int mpirank,
+                                 int nproc,
+                                 std::vector<int> &block_procs,
+                                 std::vector<int> &local_proc_blocks,
+                                 std::vector<std::vector<int> > &block_list,
+                                 std::map<std::string, char> &processed_attrs,
+                                 DecompCache &decomp_cache)
 {
-    int ierr = BP2PIO_NOERR, err_val = 0, err_cnt = 0;
+    int ierr = BP2PIO_NOERR;
     int ret = PIO_NOERR;
 
+    /* Different decompositions at different frames */
+    int  decomp_id, frame_id, fillval_exist;
+    static char decompname[PIO_MAX_NAME];
+    static char fillval_id[PIO_MAX_NAME];
+
+    std::string variable_name = varname.substr(strlen("/__pio__/var/"));
+
+    *v_base = bpIO[0].InquireVariable<T>(varname);
+    int elemsize = adios2_type_size_a2(v_base->Type());
+    assert(elemsize > 0);
+
     /* Calculate how many records/steps we have for this variable */
-    int nsteps = 1;
-    int ts = 0; /* Loop from ts to nsteps, below ts may become the last step */
+    /* Compute the number of time steps */
+    int nsteps = 0;
+    int32_t *decomp_buffer = NULL;
+    int32_t *frame_buffer = NULL;
+    char *fillval_buffer = NULL;
+    int fillval_idx = 0;
 
-    /* Compute the total number of blocks */
-    int l_nblocks = 0;
-    int g_nblocks = 0;
-    for (size_t i = 1; i <= wfiles.size(); i++)
+    try
     {
-        try
+        /* get number of application steps in adios step */
+        adios2::Variable<int> frameid_var = bpIO[0].InquireVariable<int>("/__pio__/track/frame_id/" + variable_name);
+        if (!frameid_var)
         {
-            *v_base = bpIO[i].InquireVariable<T>(varname);
-            const auto vb_blocks = bpReader[i].BlocksInfo(*v_base, 0);
-            l_nblocks += vb_blocks.size();
+            return BP2PIO_ERROR;
         }
-        catch (const std::exception &e)
+        const auto vb_blocks = bpReader[0].BlocksInfo(frameid_var, time_step);
+        for (size_t i = 0; i < vb_blocks.size(); i++)
         {
-            ierr = BP2PIO_ERROR;
-            break;
+            nsteps += vb_blocks[i].Count[0];
         }
-        catch (...)
+
+        /* read frame_id buffer */
+        frame_buffer = (int32_t*)calloc(nsteps, sizeof(int32_t));
```

Comment:
Need to handle calloc() failure case

Reply:
Yes, we should handle the calloc() failure here.